File: test/language/indexed_offsets/python/CompoundIndexedOffsetArrayTest.py
import unittest
import zserio

from testutils import getZserioApi

class CompoundIndexedOffsetArrayTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "indexed_offsets.zs").compound_indexed_offset_array

    def testBitSizeOf(self):
        createWrongOffsets = False
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        self.assertEqual(COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE, compoundIndexedOffsetArray.bitSizeOf())

    def testBitSizeOfWithPosition(self):
        createWrongOffsets = False
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        bitPosition = 1
        self.assertEqual(COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE - bitPosition,
                         compoundIndexedOffsetArray.bitSizeOf(bitPosition))

    def testInitializeOffsets(self):
        createWrongOffsets = True
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        bitPosition = 0
        self.assertEqual(COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE,
                         compoundIndexedOffsetArray.initializeOffsets(bitPosition))
        self._checkCompoundIndexedOffsetArray(compoundIndexedOffsetArray)

    def testInitializeOffsetsWithPosition(self):
        createWrongOffsets = True
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        bitPosition = 9
        self.assertEqual(COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE + bitPosition - 1,
                         compoundIndexedOffsetArray.initializeOffsets(bitPosition))

        offsetShift = 1
        self._checkOffsets(compoundIndexedOffsetArray, offsetShift)

    def testRead(self):
        writeWrongOffsets = False
        writer = zserio.BitStreamWriter()
        CompoundIndexedOffsetArrayTest._writeCompoundIndexedOffsetArrayToStream(writer, writeWrongOffsets)
        reader = zserio.BitStreamReader(writer.getByteArray(), writer.getBitPosition())
        compoundIndexedOffsetArray = self.api.CompoundIndexedOffsetArray()
        compoundIndexedOffsetArray.read(reader)
        self._checkCompoundIndexedOffsetArray(compoundIndexedOffsetArray)

    def testReadWrongOffsets(self):
        writeWrongOffsets = True
        writer = zserio.BitStreamWriter()
        CompoundIndexedOffsetArrayTest._writeCompoundIndexedOffsetArrayToStream(writer, writeWrongOffsets)
        reader = zserio.BitStreamReader(writer.getByteArray(), writer.getBitPosition())
        compoundIndexedOffsetArray = self.api.CompoundIndexedOffsetArray()
        with self.assertRaises(zserio.PythonRuntimeException):
            compoundIndexedOffsetArray.read(reader)

    def testWrite(self):
        createWrongOffsets = True
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        compoundIndexedOffsetArray.write(writer)
        self._checkCompoundIndexedOffsetArray(compoundIndexedOffsetArray)
        reader = zserio.BitStreamReader(writer.getByteArray(), writer.getBitPosition())
        readCompoundIndexedOffsetArray = self.api.CompoundIndexedOffsetArray.fromReader(reader)
        self._checkCompoundIndexedOffsetArray(readCompoundIndexedOffsetArray)
        self.assertTrue(compoundIndexedOffsetArray == readCompoundIndexedOffsetArray)

    def testWriteWithPosition(self):
        createWrongOffsets = True
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        bitPosition = 8
        writer.writeBits(0, bitPosition)
        compoundIndexedOffsetArray.write(writer)

        offsetShift = 1
        self._checkOffsets(compoundIndexedOffsetArray, offsetShift)

    def testWriteWrongOffsets(self):
        createWrongOffsets = True
        compoundIndexedOffsetArray = self._createCompoundIndexedOffsetArray(createWrongOffsets)
        writer = zserio.BitStreamWriter()
        with self.assertRaises(zserio.PythonRuntimeException):
            compoundIndexedOffsetArray.write(writer, callInitializeOffsets=False)

    @staticmethod
    def _writeCompoundIndexedOffsetArrayToStream(writer, writeWrongOffsets):
        currentOffset = ELEMENT0_OFFSET
        for i in range(NUM_ELEMENTS):
            if writeWrongOffsets and i == NUM_ELEMENTS - 1:
                writer.writeBits(WRONG_OFFSET, 32)
            else:
                writer.writeBits(currentOffset, 32)
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE

        writer.writeBits(SPACER_VALUE, 1)
        writer.writeBits(0, 7)

        for i in range(NUM_ELEMENTS):
            writer.writeBits(i, 32)
            writer.writeBits(i % 8, 3)
            if i != NUM_ELEMENTS - 1:
                writer.writeBits(0, ALIGNED_ELEMENT_SIZE - ELEMENT_SIZE)

    def _checkOffsets(self, compoundIndexedOffsetArray, offsetShift):
        offsets = compoundIndexedOffsetArray.offsets
        self.assertEqual(NUM_ELEMENTS, len(offsets))
        expectedOffset = ELEMENT0_OFFSET + offsetShift
        for offset in offsets:
            self.assertEqual(expectedOffset, offset)
            expectedOffset += ALIGNED_ELEMENT_BYTE_SIZE

    def _checkCompoundIndexedOffsetArray(self, compoundIndexedOffsetArray):
        offsetShift = 0
        self._checkOffsets(compoundIndexedOffsetArray, offsetShift)

        self.assertEqual(SPACER_VALUE, compoundIndexedOffsetArray.spacer)

        data = compoundIndexedOffsetArray.data
        self.assertEqual(NUM_ELEMENTS, len(data))
        for i in range(NUM_ELEMENTS):
            compound = data[i]
            self.assertEqual(i, compound.id)
            self.assertEqual(i % 8, compound.value)

    def _createCompoundIndexedOffsetArray(self, createWrongOffsets):
        offsets = []
        currentOffset = ELEMENT0_OFFSET
        for i in range(NUM_ELEMENTS):
            if createWrongOffsets and i == NUM_ELEMENTS - 1:
                offsets.append(WRONG_OFFSET)
            else:
                offsets.append(currentOffset)
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE

        data = []
        for i in range(NUM_ELEMENTS):
            compound = self.api.Compound(i, i % 8)
            data.append(compound)

        return self.api.CompoundIndexedOffsetArray(offsets, SPACER_VALUE, data)

NUM_ELEMENTS = 5

WRONG_OFFSET = 0

ELEMENT0_OFFSET = NUM_ELEMENTS * 4 + 1
ELEMENT_SIZE = 35
ALIGNED_ELEMENT_BYTE_SIZE = 5
ALIGNED_ELEMENT_SIZE = ALIGNED_ELEMENT_BYTE_SIZE * 8

SPACER_VALUE = 1

COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE = (NUM_ELEMENTS * 32 + 8 + (NUM_ELEMENTS - 1) * ALIGNED_ELEMENT_SIZE +
                                          ELEMENT_SIZE)

File: test/language/indexed_offsets/cpp/CompoundIndexedOffsetArrayTest.cpp
#include "gtest/gtest.h"

#include "zserio/BitStreamWriter.h"
#include "zserio/BitStreamReader.h"

#include "indexed_offsets/compound_indexed_offset_array/CompoundIndexedOffsetArray.h"

namespace indexed_offsets
{
namespace compound_indexed_offset_array
{

class CompoundIndexedOffsetArrayTest : public ::testing::Test
{
protected:
    void writeCompoundIndexedOffsetArrayToByteArray(zserio::BitStreamWriter& writer, bool writeWrongOffsets)
    {
        const uint32_t wrongOffset = WRONG_OFFSET;
        uint32_t currentOffset = ELEMENT0_OFFSET;
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            if ((i + 1) == NUM_ELEMENTS && writeWrongOffsets)
                writer.writeBits(wrongOffset, 32);
            else
                writer.writeBits(currentOffset, 32);
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE;
        }

        writer.writeBits(SPACER_VALUE, 1);
        writer.writeBits(0, 7);

        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            writer.writeBits(i, 32);
            writer.writeBits(i % 8, 3);
            if ((i + 1) != NUM_ELEMENTS)
                writer.writeBits(0, ALIGNED_ELEMENT_SIZE - ELEMENT_SIZE);
        }
    }

    void checkOffsets(const CompoundIndexedOffsetArray& compoundIndexedOffsetArray, uint16_t offsetShift)
    {
        const auto& offsets = compoundIndexedOffsetArray.getOffsets();
        const size_t expectedNumElements = NUM_ELEMENTS;
        ASSERT_EQ(expectedNumElements, offsets.size());
        uint32_t expectedOffset = ELEMENT0_OFFSET + offsetShift;
        for (auto it = offsets.begin(); it != offsets.end(); ++it)
        {
            ASSERT_EQ(expectedOffset, *it);
            expectedOffset += ALIGNED_ELEMENT_BYTE_SIZE;
        }
    }

    void checkCompoundIndexedOffsetArray(const CompoundIndexedOffsetArray& compoundIndexedOffsetArray)
    {
        const uint16_t offsetShift = 0;
        checkOffsets(compoundIndexedOffsetArray, offsetShift);

        const uint8_t expectedSpacer = SPACER_VALUE;
        ASSERT_EQ(expectedSpacer, compoundIndexedOffsetArray.getSpacer());

        const auto& data = compoundIndexedOffsetArray.getData();
        const size_t expectedNumElements = NUM_ELEMENTS;
        ASSERT_EQ(expectedNumElements, data.size());
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            const Compound& compound = data[i];
            ASSERT_EQ(i, compound.getId());
            ASSERT_EQ(i % 8, compound.getValue());
        }
    }

    void fillCompoundIndexedOffsetArray(CompoundIndexedOffsetArray& compoundIndexedOffsetArray,
            bool createWrongOffsets)
    {
        auto& offsets = compoundIndexedOffsetArray.getOffsets();
        offsets.reserve(NUM_ELEMENTS);
        const uint32_t wrongOffset = WRONG_OFFSET;
        uint32_t currentOffset = ELEMENT0_OFFSET;
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            if ((i + 1) == NUM_ELEMENTS && createWrongOffsets)
                offsets.push_back(wrongOffset);
            else
                offsets.push_back(currentOffset);
            currentOffset += ALIGNED_ELEMENT_BYTE_SIZE;
        }
        compoundIndexedOffsetArray.setSpacer(SPACER_VALUE);

        auto& data = compoundIndexedOffsetArray.getData();
        data.reserve(NUM_ELEMENTS);
        for (uint8_t i = 0; i < NUM_ELEMENTS; ++i)
        {
            Compound compound;
            compound.setId(i);
            compound.setValue(i % 8);
            data.push_back(compound);
        }
    }

    static const uint8_t    NUM_ELEMENTS = 5;

    static const uint32_t   WRONG_OFFSET = 0;

    static const uint32_t   ELEMENT0_OFFSET = NUM_ELEMENTS * sizeof(uint32_t) + sizeof(uint8_t);
    static const uint8_t    ELEMENT_SIZE = 35;
    static const uint8_t    ALIGNED_ELEMENT_SIZE = 5 * 8;
    static const uint8_t    ALIGNED_ELEMENT_BYTE_SIZE = ALIGNED_ELEMENT_SIZE / 8;

    static const uint8_t    SPACER_VALUE = 1;

    static const size_t     COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE = ELEMENT0_OFFSET * 8 +
            (NUM_ELEMENTS - 1) * ALIGNED_ELEMENT_SIZE + ELEMENT_SIZE;

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
};

TEST_F(CompoundIndexedOffsetArrayTest, read)
{
    const bool writeWrongOffsets = false;
    zserio::BitStreamWriter writer(bitBuffer);
    writeCompoundIndexedOffsetArrayToByteArray(writer, writeWrongOffsets);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    CompoundIndexedOffsetArray compoundIndexedOffsetArray(reader);
    checkCompoundIndexedOffsetArray(compoundIndexedOffsetArray);
}

TEST_F(CompoundIndexedOffsetArrayTest, readWrongOffsets)
{
    const bool writeWrongOffsets = true;
    zserio::BitStreamWriter writer(bitBuffer);
    writeCompoundIndexedOffsetArrayToByteArray(writer, writeWrongOffsets);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    EXPECT_THROW(CompoundIndexedOffsetArray compoundIndexedOffsetArray(reader),
            zserio::CppRuntimeException);
}

TEST_F(CompoundIndexedOffsetArrayTest, bitSizeOf)
{
    const bool createWrongOffsets = false;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    const size_t expectedBitSize = COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, compoundIndexedOffsetArray.bitSizeOf());
}

TEST_F(CompoundIndexedOffsetArrayTest, bitSizeOfWithPosition)
{
    const bool createWrongOffsets = false;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 1;
    const size_t expectedBitSize = COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE - bitPosition;
    ASSERT_EQ(expectedBitSize, compoundIndexedOffsetArray.bitSizeOf(bitPosition));
}

TEST_F(CompoundIndexedOffsetArrayTest, initializeOffsets)
{
    const bool createWrongOffsets = true;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 0;
    const size_t expectedBitSize = COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE;
    ASSERT_EQ(expectedBitSize, compoundIndexedOffsetArray.initializeOffsets(bitPosition));
    checkCompoundIndexedOffsetArray(compoundIndexedOffsetArray);
}

TEST_F(CompoundIndexedOffsetArrayTest, initializeOffsetsWithPosition)
{
    const bool createWrongOffsets = true;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 9;
    const size_t expectedBitSize = COMPOUND_INDEXED_OFFSET_ARRAY_BIT_SIZE + bitPosition - 1;
    ASSERT_EQ(expectedBitSize, compoundIndexedOffsetArray.initializeOffsets(bitPosition));

    const uint16_t offsetShift = 1;
    checkOffsets(compoundIndexedOffsetArray, offsetShift);
}

TEST_F(CompoundIndexedOffsetArrayTest, write)
{
    const bool createWrongOffsets = false;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    zserio::BitStreamWriter writer(bitBuffer);
    compoundIndexedOffsetArray.write(writer);
    checkCompoundIndexedOffsetArray(compoundIndexedOffsetArray);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    CompoundIndexedOffsetArray readCompoundIndexedOffsetArray(reader);
    checkCompoundIndexedOffsetArray(readCompoundIndexedOffsetArray);
    ASSERT_TRUE(compoundIndexedOffsetArray == readCompoundIndexedOffsetArray);
}

TEST_F(CompoundIndexedOffsetArrayTest, writeWithPosition)
{
    const bool createWrongOffsets = true;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    const size_t bitPosition = 8;
    compoundIndexedOffsetArray.initializeOffsets(bitPosition);
    zserio::BitStreamWriter writer(bitBuffer);
    writer.writeBits(0, bitPosition);
    compoundIndexedOffsetArray.write(writer);

    const uint16_t offsetShift = 1;
    checkOffsets(compoundIndexedOffsetArray, offsetShift);
}

TEST_F(CompoundIndexedOffsetArrayTest, writeWrongOffsets)
{
    const bool createWrongOffsets = true;
    CompoundIndexedOffsetArray compoundIndexedOffsetArray;
    fillCompoundIndexedOffsetArray(compoundIndexedOffsetArray, createWrongOffsets);

    zserio::BitStreamWriter writer(bitBuffer);
    ASSERT_THROW(compoundIndexedOffsetArray.write(writer), zserio::CppRuntimeException);
}

} // namespace compound_indexed_offset_array
} // namespace indexed_offsets
